Add GDI Covert Operations - Twist of Fate - scg41ea

YAmismo opened the pull request:
Same from #20743.
Sorry i messed up all. Still need to learn more git.

YAmismo committed:
Add GDI Covert Operations - Twist of Fate - scg41ea

JovialFeline commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,263 @@
+		end)
+		Trigger.OnAllKilled(NodSams, function()
+			AirSupport = Actor.Create("Astk.proxy", true, { Owner = Player1 })
+			Player1.MarkCompletedObjective(DestroySams)
```
I can sympathize with the git goof. Not sure if you made other changes but it looks like the last crash fix was reverted.
```suggestion
		if DestroySams then
			Player1.MarkCompletedObjective(DestroySams)
		end
```

YAmismo commented:
Fixed a few posible objective crashes.

PunkPun commented:
Needs a rebase and difficulty translations added. See changelog #20585 for how.

YAmismo commented:
Rebased and translations added.

PunkPun commented on mods/cnc/maps/twist-of-fate/rules.yaml:
```diff
@@ -0,0 +1,113 @@
+		StartVideo: airstrk.vqa
+		LossVideo: obel.vqa
+		WinVideo: paratrop.vqa
+		Briefing: We had set up a small recon post in NOD territory, but they captured and reinforced it.\n\nGetting that post back would be a major coup.\n\nAn MCV and armed convoy are on their way to aid you in establishing a new base.\n\nOnce established, eliminate all NOD forces in the area.
```
```suggestion
		Briefing: We had set up a small recon post in Nod territory, but they captured and reinforced it.\n\nGetting that post back would be a major coup.\n\nAn MCV and armed convoy are on their way to aid you in establishing a new base.\n\nOnce established, eliminate all Nod forces in the area.
```

PunkPun commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,274 @@
+		Trigger.AfterDelay(DateTime.Seconds(20), function()
+			if CaptureCommCenter == nil then
+				Media.DisplayMessage(UserInterface.Translate("air-strikes-intel-report"))
+				CaptureCommCenter = AddSecondaryObjective(Player1, "capture-nod-communications-center")
```
Could we add a delay between messages? There's a lot of text

PunkPun reviewed (commented):
Would it make sense to add periodic reinforcements on easy difficulty? The Mission is quite lengthy

YAmismo commented:
Done.
Periodic reinforcements on easy difficulty? What could go wrong :sunglasses:.

PunkPun commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,306 @@
+
+	Trigger.OnAllKilled(NodSams, function()
+		if DestroySams == nil then
+			DestroySams = AddSecondaryObjective(Player1, "destroy-sams")
```
if the first secondary objective was failed and this objective wasn't even announced, IMO it makes little sense to still mention it to the player. We also shouldn't be adding the support power

YAmismo commented:
Modified sams objective.

PunkPun reviewed (commented):
I noticed all units aren't coloured. They are all of blue color, while structures & harvesters are red. Is this intentional?

The game crashed when I captured the radar.
```Attempted to get trait from destroyed object (fact 77 (not in world))```
 
Somehow the chinooks didn't land. They just stood there. I'd suggest just adding an additional reinforcement rather than making the existing one bigger
<img width="228" alt="Screenshot 2023-05-07 at 16 21 31" src="https://user-images.githubusercontent.com/37534529/236680407-b8282fcd-31e4-49eb-a39b-3ec4c758ef98.png">

PunkPun commented:
Also got a crash if I captured the coms sentre after already having killed sams

```Attempted to get trait from destroyed object (sam 101 (not in world))```

abcdefg30 commented:
> I noticed all units aren't coloured. They are all of blue color, while structures & harvesters are red. Is this intentional?

This is probably just the default campaign coloring? (See the shared `campaign-palettes.yaml` file.)

YAmismo commented:
Fixed both crashes from comm center capture and only 1 chinook should come now (was because of entering the cell trigger with multiple infantry at the same time).

About the colors... i'm with abcdefg30, just imported the map with the utility and added cnc|rules/campaign-palettes.yaml to map.yaml.

PunkPun reviewed (commented):
Now when I capture the coms centre before the objective appears, I don't get the objective to kill sams.

I still managed to get the crash somehow`Attempted to get trait from destroyed object (sam 101 (not in world)`

YAmismo commented:
> Now when I capture the coms centre before the objective appears, I don't get the objective to kill sams.

It's intended, before the first airstrike it is just a standard comm center, nothing special, i guess.

Fixed that crash too, my bad, stupid copy-paste xd.

PunkPun reviewed (commented):
It now crashes when I kill the conyard.
`Fatal Lua Error: [string "twist-of-fate.lua"]:83: attempt to index global 'NodAirSupport' (a nil value)`

YAmismo commented:
Fixed.

PunkPun reviewed (dismissed):
LGTM

abcdefg30 commented on mods/cnc/languages/lua/en.ftl:
```diff
@@ -150,3 +150,11 @@ nod-soldier = Nod Soldier
+## twist-of-fate
+clear-path = Repel the ambush and clear the way
+    for our MCV.
+air-strikes-intel-report = Nod airstrikes are being directed by a communications center, located northwest. We may gain useful information from its capture.
```
```suggestion
air-strikes-intel-report = Nod airstrikes are being directed by a Communications Center, located northwest. We may gain useful information from its capture.
```
to be consistent with the objective text.

abcdefg30 commented on mods/cnc/maps/twist-of-fate/map.yaml:
```diff
@@ -0,0 +1,1124 @@
+		Color: F5D378
+		LockSpawn: True
+		LockTeam: True
+		Allies: GDI
```
Why is this player allied to itself? (Same for Nod.)

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,327 @@
+WorldLoaded = function()
+
+	Camera.Position = DefaultCameraPosition.CenterPosition
+	Player1 = Player.GetPlayer("GDI")
```
This should be `GDI = Player.GetPlayer("GDI")` to match the other scripts. (Also see #20698.)

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,327 @@
+
+	Camera.Position = DefaultCameraPosition.CenterPosition
+	Player1 = Player.GetPlayer("GDI")
+	EnemyAi = Player.GetPlayer("Nod")
```
This should be `Nod = Player.GetPlayer("Nod")` to match the other scripts. (Also see #20698.)

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,327 @@
+						return
+					end
+					if NodAstkHq.Owner == Player1 then
+						Trigger.AfterDelay(100, CompleteCaptureCommCenterObjective)
```
```suggestion
						Trigger.AfterDelay(DateTime.Seconds(4), CompleteCaptureCommCenterObjective)
```

abcdefg30 commented on mods/cnc/languages/lua/en.ftl:
```diff
@@ -150,3 +150,11 @@ nod-soldier = Nod Soldier
+air-strikes-intel-report = Nod airstrikes are being directed by a communications center, located northwest. We may gain useful information from its capture.
+capture-nod-communications-center = Capture the Nod Communications Center to
+    the northwest.
+communications-center-captured-sams-located = Our engineers located Nod SAM sites. They also disabled an unusual trap or alarm on a construction yard to the south.
```
"trap or alarm" sounds a bit odd. Can we just name this something like
```suggestion
communications-center-captured-sams-located = Our engineers located Nod SAM sites. They also disarmed an unusual trap on the Construction Yard to the south.
```
?

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,327 @@
+		AmbushTeam = Map.ActorsInBox(Map.CenterOfCell(CPos.New(46,5)), Map.CenterOfCell(CPos.New(60,53)), function(a) return a.Owner == EnemyAi and a.Type ~= "stnk" end)
+		Trigger.OnAllKilled(AmbushTeam, function()
+			Player1.MarkCompletedObjective(ClearPath)
+			Trigger.AfterDelay(50, function()
+				Trigger.AfterDelay(750, function()
+					Flare.Destroy()
+				end)
```
```suggestion
			Trigger.AfterDelay(DateTime.Seconds(2), function()
				Trigger.AfterDelay(DateTime.Seconds(30), Flare.Destroy)
```
should also work.

YAmismo commented:
Done.

abcdefg30 reviewed (commented):
The mission play well, good job! I find the secondary objective of destroying the SAM Sites a bit odd. I don't see a scenario where a player destroys the SAMs before having most of the main base destroyed already, making it a bit pointless. The briefing mentions that `Getting that post back would be a major coup`, so we could make the second objective about capturing the middle Construction Yard instead and always give the player an airstrike when they have a Comm Center. What do you think?

abcdefg30 commented on mods/cnc/maps/twist-of-fate/rules.yaml:
```diff
@@ -0,0 +1,113 @@
+		Range: 7c0
+
+A10:
+	Targetable:
```
This needs to be
```suggestion
	Targetable:
		TargetTypes: Air
```
for it to work. The planes are way too faster for all anti-air though.

YAmismo commented:
> I find the secondary objective of destroying the SAM Sites a bit odd. I don't see a scenario where a player destroys the SAMs before having most of the main base destroyed already, making it a bit pointless. The briefing mentions that Getting that post back would be a major coup, so we could make the second objective about capturing the middle Construction Yard instead and always give the player an airstrike when they have a Comm Center. What do you think?

Yes, i thought about this while coding the map... not sure why i didn't do it.
You mean, remove sams objective but still have both secondary objectives, capture GDI old Con Yard and capture Comm Center too for air support?

abcdefg30 commented:
> You mean, remove sams objective but still have both secondary objectives, capture GDI old Con Yard and capture Comm Center too for air support?

Basically, but I would leave the Comm Center objective as is. Tying airstrikes to capturing the old Con Yard feels a bit weird. I'd either let the player have airstrikes like in skirmish or disable them entirely for this mission if they are too powerful.

PunkPun commented:
When it comes to missions, I don't think you can make anything too powerfull

YAmismo commented:
![imagen](https://github.com/OpenRA/OpenRA/assets/30934768/33b1e9b8-6469-483f-9913-bb2e0d02559a)

Could this be interesting? It was a GDI base we have to recover.
Or is it too much change to the original mission?

abcdefg30 commented:
Looks good to me. 👍

YAmismo commented:
Added recover recon post secondary objective.
Removed destroy sams secondary objective.

abcdefg30 commented on mods/cnc/languages/lua/en.ftl:
```diff
@@ -13,7 +13,7 @@ destroy-obelisk-sams = Destroy the SAM sites protecting the Obelisk.
 destroy-biotech = Destroy the biotech facility.
 
-## gdi03, gdi05abc, gdi07 gdi08ab, gdi09
+## gdi03, gdi05abc, gdi07 gdi08ab, gdi09, twist-of-fate
```
This can be removed again now.

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate-AI.lua:
```diff
@@ -0,0 +1,647 @@
+
+SP1 = { waypoint7.Location, CPos.New(14,29), CPos.New(15,15), CPos.New(49,14), CPos.New(60,38), waypoint4.Location }
+SP2 = { waypoint7.Location, CPos.New(14,29), CPos.New(15,15), CPos.New(27,13), waypoint9.Location }
+SneakPahts = { SP1 }
```
```suggestion
SneakPaths = { SP1 }
```

abcdefg30 commented on mods/cnc/languages/lua/en.ftl:
```diff
@@ -150,3 +150,13 @@ nod-soldier = Nod Soldier
+air-strikes-intel-report = Nod airstrikes are being directed by a Communications Center, located northwest. We may gain useful information from its capture.
+capture-nod-communications-center = Capture the Nod Communications Center to
+    the northwest.
+communications-center-captured-sams-located = Our engineers located Nod SAM sites. They also disarmed an unusual trap on the Construction Yard to the south.
```
I guess we should remove the SAM location revealing too now.
```suggestion
communications-center-captured = Communications Center captured. Our engineers managed to disarm an unusual trap on the Construction Yard to the south.
```
and remove the SAM revealing logic?

PunkPun replied:
I think it's fine to keep it

abcdefg30 replied:
Hm, I guess we could keep it so that the player can maneuver their Orcas a bit more freely, but I still think it is weird.

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,328 @@
+
+	Trigger.AfterDelay(1,function()
+		AmbushTeam = Map.ActorsInBox(Map.CenterOfCell(CPos.New(46,5)), Map.CenterOfCell(CPos.New(60,53)), function(a) return a.Owner == Nod and a.Type ~= "stnk" end)
+		local neutralHarv = Player.GetPlayer("Neutral").GetActorsByType("harv")
+		neutralHarv[1].Destroy()
```
I assume that destroys the harvester spawned by the refinery. This should instead be done by specifying `FreeActor: false` on the actor definition in map.yaml, see https://github.com/OpenRA/OpenRA/blob/621c85059ec92f550ad25044063f73ed326ae8fa/mods/cnc/maps/nod07a/map.yaml#L538-L541 for an example.

YAmismo commented:
Fixed.

abcdefg30 commented on mods/cnc/maps/twist-of-fate/rules.yaml:
```diff
@@ -0,0 +1,110 @@
+		Range: 7c0
+
+HQ:
+	-AirstrikePower:
```
Can we re-enable this now?

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate-AI.lua:
```diff
@@ -0,0 +1,647 @@
+	end
+	if #ProductionQueue["vehicle"] > 0 and not ProductionBuildings["vehicle"].IsDead and not ProductionBuildings["vehicle"].IsProducing("harv") and not WaitingAirDrop then
+		ProductionBuildings["vehicle"].Build({ ProductionQueue["vehicle"][1] }, function(u)
+			table.insert(NewTeam, u[1])
```
I just got a weird crash on my latest play through:
<details>

```
OpenRA engine version {DEV_VERSION}
OpenRA Language: en
Tiberian Dawn mod version {DEV_VERSION}
on map 5c165ed48eac6ce4b4288262f8df3304ee176ea7 (GDI - Twist of Fate by Westwood Studios).
Date: 2023-05-18 19:53:44Z
Operating System: Windows (X64, Microsoft Windows NT 10.0.19045.0)
Runtime Version: .NET CLR 6.0.16
Installed Language: en (Installed) en (Current) en (Current UI)
Exception of type `Eluant.LuaException`: [string "-- Make copies of global functions to preve..."]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Eluant.LuaException: [string "-- Make copies of global functions to preve..."]:30: Actor 'harv' does not define a property 'AttackMove'
   at Eluant.LuaRuntime.Call(IList`1 args)
   at Eluant.LuaRuntime.Call(LuaFunction fn, IList`1 args)
   at OpenRA.Mods.Common.Scripting.UtilsGlobal.Do(LuaValue[] collection, LuaFunction func) in \OpenRA\OpenRA.Mods.Common\Scripting\Global\UtilsGlobal.cs:line 29
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OpenRA.Scripting.ScriptMemberWrapper.Invoke(LuaVararg args) in \OpenRA\OpenRA.Game\Scripting\ScriptMemberWrapper.cs:line 76
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Eluant.LuaRuntime.MethodWrapper.Invoke(Object[] parms)
   at Eluant.LuaRuntime.MakeManagedCall(IntPtr state, MethodWrapper wrapper)
   at Eluant.LuaRuntime.Call(IList`1 args)
   at Eluant.LuaRuntime.Call(LuaFunction fn, IList`1 args)
   at OpenRA.Mods.Common.Scripting.ProductionQueueProperties.<>c__DisplayClass3_2.<Build>b__4(Actor factory, Actor unit) in \OpenRA\OpenRA.Mods.Common\Scripting\Properties\ProductionProperties.cs:line 172
   at OpenRA.Mods.Common.Scripting.ScriptTriggers.OpenRA.Mods.Common.Traits.INotifyProduction.UnitProduced(Actor self, Actor other, CPos exit) in \OpenRA\OpenRA.Mods.Common\Scripting\ScriptTriggers.cs:line 177
   at OpenRA.Mods.Common.Traits.Production.<>c__DisplayClass6_0.<DoProduction>b__0(World w) in \OpenRA\OpenRA.Mods.Common\Traits\Production.cs:line 99
   at OpenRA.World.Tick() in \OpenRA\OpenRA.Game\World.cs:line 453
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in \OpenRA\OpenRA.Game\Game.cs:line 630
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:line 645
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:line 814
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:line 867
   at OpenRA.Game.InitializeAndRun(String[] args) in \OpenRA\OpenRA.Game\Game.cs:line 305
   at OpenRA.Launcher.Program.Main(String[] args) in \OpenRA\OpenRA.Launcher\Program.cs:line 32
```
</details>

The Lua `Build` function has always been a bit buggy, so it's not out of the question that it somehow built a harvester and ran into the callback for normal vehicle production. (I don't know how else to explain a Harvester trying to attack move. Or do you have another idea?) Can we add a guard against that here? Something like
```suggestion
			if u[1].Type == "harv" then
				return
			end

			table.insert(NewTeam, u[1])
```

YAmismo replied:
That's why i added a delay between nod vehicles production. Not sure how to explain. The thing is... if one unit is ready, and a second one is queued before the plane arrives and leaves it on the floor, weird things happens... It's like the first unit doesn't execute his function and the next unit executes both, in this case the harvester try to attackmove.

abcdefg30 commented on mods/cnc/maps/twist-of-fate/twist-of-fate.lua:
```diff
@@ -0,0 +1,326 @@
+
+	Trigger.OnKilled(NodOutPostCYard, function()
+		SendNodAirstrike(false)
+		GDI.MarkFailedObjective(RecoverOldBase)
```
This should check if the objective exists before marking it as failed (you can destroy the Conyard before the reinforcements arrive.)
```suggestion
		if not RecoverOldBase then
			RecoverOldBase = AddSecondaryObjective(GDI, "recover-old-base")
		end

		GDI.MarkFailedObjective(RecoverOldBase)
```

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c684cfd27957230c4132a1513fa2fd45b71dbf46...f31b0b13a2f12dcd4af92aa88a44680d9477b2f0)
